{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8593d81e",
   "metadata": {},
   "source": [
    "## ae expanded & annotate network\n",
    "use pfib dataset to run aberrant network analysis 20250520"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f3d5527a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('default')\n",
    "\n",
    "import os, sys\n",
    "from tqdm import tqdm\n",
    "import importlib\n",
    "import logging\n",
    "import datetime\n",
    "\n",
    "# load ae_net source code\n",
    "sourcedir = '/mnt/disk7t/xwj/axolotl_rev/gitcode'\n",
    "sys.path.append(sourcedir)\n",
    "# from functions import *\n",
    "# %run functions.py\n",
    "import functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "df6844a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'functions' from '/mnt/disk7t/xwj/axolotl_rev/gitcode/functions.py'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importlib.reload(functions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0e043a8",
   "metadata": {},
   "source": [
    "load packaged functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed1918c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# output directory\n",
    "outdir = '/mnt/disk7t/xwj/axolotl_rev/output' \n",
    "os.makedirs(outdir+'/log', exist_ok=True)\n",
    "os.makedirs(outdir+'/data', exist_ok=True)\n",
    "os.makedirs(outdir+'/sweet', exist_ok=True)\n",
    "\n",
    "# Defining Filename Templates:\n",
    "filename_nd_template = '{outdir}/data/node_df_{sample}.txt'\n",
    "filename_ed_template = '{outdir}/data/edge_df_{sample}.txt'\n",
    "filename_ed_aberrant_template = '{outdir}/data/edge_df_{sample}_aberrant.txt'\n",
    "filename_scatter_template = '{outdir}/data/edge_scatter_{sample}.pdf'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "850c3e07",
   "metadata": {},
   "source": [
    "reading and processing different datasets related to gene expression and outlier detection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "47c0da66",
   "metadata": {},
   "outputs": [],
   "source": [
    "## pfib423 dataset\n",
    "# path to files from upstream AE detection tools\n",
    "# your_dataset_file = '/media/eys/xwj/axolotl_rev/pfib_423/df_cts_gm2022_dedup.txt'\n",
    "# datadir = '/mnt/disk7t/xwj/axolotl_rev/pfib_423/'\n",
    "datadir = f'{sourcedir}/example'\n",
    "result_baseline = {\n",
    "    'cts': f'{datadir}/t0_g12543_s423.txt.gz',\n",
    "    'outsingle':f'{datadir}/osg_res_1_001.txt',\n",
    "    # 'outrider': f'{datadir}/outrider_res_1_001.txt',\n",
    "    # 'outrider_cts': f'{datadir}/outrider_norm_cts_1_001.txt',\n",
    "    'trueoutlier': f'{datadir}/df_true49.txt',\n",
    "    'sampleanno': f'{datadir}/sample_annot.tsv', # all samples meta information，useful in batch removal\n",
    "}\n",
    "#------------------------------------------------------------------------#\n",
    "file = result_baseline['sampleanno']\n",
    "s_anno = pd.read_csv(file, sep=\"\\t\", index_col='RNA_ID')\n",
    "# pfib423 have two types sequencing library: strand & non strand. split as two groups\n",
    "# rename: stranded -> ss； non-stranded -> ns\n",
    "s_anno['group'] = s_anno['STRAND_SPECIFIC'].map({True:'ss', False:'ns'})\n",
    "s_anno.describe()\n",
    "\n",
    "#------------------------------------------------------------------------#\n",
    "# results of upstream tools\n",
    "# method 1. outsingle\n",
    "file = result_baseline['outsingle']\n",
    "score = pd.read_csv(file, sep=\"\\t\", index_col=0)\n",
    "\n",
    "if False:\n",
    "    # method 2. if use OUTRIDER res\n",
    "    file = result_baseline['outrider']\n",
    "    score = pd.read_csv(file, sep=\"\\t\", index_col=0)\n",
    "    score = score.pivot_table(values='pValue',index= 'geneID', columns='sampleID')\n",
    "    cts_outrider = pd.read_csv(result_baseline['outrider_cts'], sep=\"\\t\", index_col=0)\n",
    "\n",
    "#------------------------------------------------------------------------#\n",
    "# known outliers\n",
    "outlier = pd.read_csv(result_baseline['trueoutlier'], sep=\"\\t\", index_col=None)\n",
    "outlier['group'] = s_anno.loc[ outlier['Sample'],'group'].values\n",
    "\n",
    "#------------------------------------------------------------------------#\n",
    "# read count table\n",
    "file = result_baseline['cts']\n",
    "cts = pd.read_csv(file, sep=\"\\t\", index_col=0)\n",
    "exp = cts.transform(lambda x: np.log(x+1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ba561d0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Gene</th>\n",
       "      <th>RNAdefects</th>\n",
       "      <th>TableName</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R80184</td>\n",
       "      <td>M</td>\n",
       "      <td>ALDH18A1</td>\n",
       "      <td>AE, MAE</td>\n",
       "      <td>S2</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R60537</td>\n",
       "      <td>M</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>AE, Var</td>\n",
       "      <td>S2</td>\n",
       "      <td>ss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R62943</td>\n",
       "      <td>F</td>\n",
       "      <td>MICOS13</td>\n",
       "      <td>AE, AS</td>\n",
       "      <td>S2</td>\n",
       "      <td>ss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R96820</td>\n",
       "      <td>F</td>\n",
       "      <td>CLPP</td>\n",
       "      <td>AE, AS</td>\n",
       "      <td>S2</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R77611</td>\n",
       "      <td>F</td>\n",
       "      <td>DLD</td>\n",
       "      <td>AE, MAE</td>\n",
       "      <td>S2</td>\n",
       "      <td>ss</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample Sex      Gene RNAdefects TableName group\n",
       "0  R80184   M  ALDH18A1    AE, MAE        S2    ns\n",
       "1  R60537   M   ATP6AP1    AE, Var        S2    ss\n",
       "2  R62943   F   MICOS13     AE, AS        S2    ss\n",
       "3  R96820   F      CLPP     AE, AS        S2    ns\n",
       "4  R77611   F       DLD    AE, MAE        S2    ss"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outlier.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6d1b6152",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDIVIDUAL_ID</th>\n",
       "      <th>TISSUE</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AFFECTED</th>\n",
       "      <th>ICD_10</th>\n",
       "      <th>PAIRED_END</th>\n",
       "      <th>STRAND_SPECIFIC</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R0052</th>\n",
       "      <td>p11688</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Female</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R0053</th>\n",
       "      <td>p48517</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Male</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R19907</th>\n",
       "      <td>p46808</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Male</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R10581</th>\n",
       "      <td>p51250</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Female</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R0055</th>\n",
       "      <td>p5413</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Female</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       INDIVIDUAL_ID      TISSUE     SEX  AFFECTED ICD_10  PAIRED_END  \\\n",
       "RNA_ID                                                                  \n",
       "R0052         p11688  Fibroblast  Female      True    E88        True   \n",
       "R0053         p48517  Fibroblast    Male      True    E88        True   \n",
       "R19907        p46808  Fibroblast    Male      True    E88        True   \n",
       "R10581        p51250  Fibroblast  Female      True    E88        True   \n",
       "R0055          p5413  Fibroblast  Female      True    E88        True   \n",
       "\n",
       "        STRAND_SPECIFIC group  \n",
       "RNA_ID                         \n",
       "R0052             False    ns  \n",
       "R0053             False    ns  \n",
       "R19907            False    ns  \n",
       "R10581            False    ns  \n",
       "R0055             False    ns  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_anno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c658dd1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ss 269 25\n",
      "ns 154 24\n",
      "{'ss': {'exp': '/mnt/disk7t/xwj/axolotl_rev/output/df_exp_ss269.txt', 'score': '/mnt/disk7t/xwj/axolotl_rev/output/df_score_ss269.txt', 'outlier': '/mnt/disk7t/xwj/axolotl_rev/output/df_true_ss25.txt'}, 'ns': {'exp': '/mnt/disk7t/xwj/axolotl_rev/output/df_exp_ns154.txt', 'score': '/mnt/disk7t/xwj/axolotl_rev/output/df_score_ns154.txt', 'outlier': '/mnt/disk7t/xwj/axolotl_rev/output/df_true_ns24.txt'}}\n"
     ]
    }
   ],
   "source": [
    "# ------------------------------\n",
    "#  Dataset preprocess\n",
    "# ------------------------------\n",
    "# dataset split by sequencing library type (stranded/non-stranded) as ss/ns batches\n",
    "batches = ['ss','ns']\n",
    "\n",
    "# filename dict\n",
    "dict_filename = dict.fromkeys(batches)\n",
    "\n",
    "for group in batches:\n",
    "    \n",
    "    # expression matrix \n",
    "    sample_list = s_anno.query('group == @group').index.tolist()\n",
    "    expfile = f'{outdir}/df_exp_{group}{len(sample_list)}.txt'\n",
    "    exp.loc[:, sample_list].to_csv( expfile, sep='\\t')\n",
    "    \n",
    "    # score matrix\n",
    "    scorefile = f'{outdir}/df_score_{group}{len(sample_list)}.txt'\n",
    "    score.loc[:, sample_list].to_csv( scorefile, sep='\\t')\n",
    "    \n",
    "    # sample meta table\n",
    "    sample_list_interest = outlier.query('group == @group')['Sample']\n",
    "    outlierfile = f'{outdir}/df_true_{group}{len(sample_list_interest)}.txt'\n",
    "    outlier.query('Sample in @sample_list_interest').to_csv(outlierfile,sep='\\t')\n",
    "        \n",
    "    dict_filename[group] = {}\n",
    "    dict_filename[group]['exp'] = expfile\n",
    "    dict_filename[group]['score'] = scorefile\n",
    "    dict_filename[group]['outlier'] = outlierfile\n",
    "    print(group, len(sample_list), len(sample_list_interest))\n",
    "    \n",
    "print(dict_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b9282155",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------------------------------\n",
    "#  Sample Selection \n",
    "# ------------------------------\n",
    "\n",
    "# s_list = [ 'R52016','R28774'] #from ns sample\n",
    "# s_list = [ \"R15264\" ] #from ss sample\n",
    "\n",
    "# manually choose one group to run. \n",
    "group = 'ns' \n",
    "# load data\n",
    "file_e = dict_filename[group]['exp']\n",
    "df_exp = pd.read_csv(file_e, sep=\"\\t\", index_col=0)\n",
    "\n",
    "file_sc = dict_filename[group]['score']\n",
    "df_score = pd.read_csv(file_sc, sep=\"\\t\", index_col=0)\n",
    "\n",
    "outlierfile = dict_filename[group]['outlier']\n",
    "df_outlier = pd.read_csv(outlierfile, sep=\"\\t\", index_col=0)\n",
    "\n",
    "# select subset of samples with outliers\n",
    "s_list = df_outlier['Sample'].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a331c3f4",
   "metadata": {},
   "source": [
    "run aberrant network and edge analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2d330dcd-6895-4003-afd3-37d49e1a414a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reload functions in case of code changes\n",
    "importlib.reload(functions)\n",
    "# current time format as string for log file name\n",
    "current_time = datetime.datetime.now().strftime(\"%Y%m%d_%H%M%S\")\n",
    "\n",
    "logging.basicConfig(filename=f'{outdir}/log/ae_net_{current_time}.log', level=logging.INFO, \n",
    "                format='%(asctime)s:%(levelname)s:%(message)s')\n",
    "\n",
    "#-----------------------------------------------------------\n",
    "# prepare intermediate data\n",
    "#-----------------------------------------------------------\n",
    "\n",
    "# correlation\n",
    "pcc = pd.DataFrame( np.corrcoef(df_exp), index=df_exp.index.tolist(), columns=df_exp.index.tolist())\n",
    "pcc.index.name, pcc.columns.name ='gene1','gene2'\n",
    "logging.info(\"Prepare coexpression pcc matrix\")\n",
    "\n",
    "# pcc remove duplicate, get undirected network = 0.5 min\n",
    "# pcc = dict_pcc[c][ctype][p].copy()\n",
    "# pcc_tri = pd.Series(pcc.stack()).reset_index().rename(columns={0:'pcc'})\n",
    "# pcc_tri = pcc_tri[ pcc_tri['gene1'] < pcc_tri['gene2'] ]\n",
    "\n",
    "pcc_tri = functions.pearson_extract_upper_triangle(pcc)\n",
    "logging.info(\"Prepare coexpression edge table\")\n",
    "\n",
    "logging.info(f\"{'DataFrame':<15} | {'Shape':<10}\")\n",
    "logging.info(\"-\" * 40)\n",
    "logging.info(f\"{'df_outlier':<15} | {df_outlier.shape}\")\n",
    "logging.info(f\"{'df_exp':<15} | {df_exp.shape}\")\n",
    "logging.info(f\"{'df_score':<15} | {df_score.shape}\")\n",
    "logging.info(f\"{'pcc':<15} | {pcc.shape}\")\n",
    "logging.info(f\"{'pcc_tri':<15} | {pcc_tri.shape}\")\n",
    "logging.info(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e159860f-bbd8-4bde-ac58-a935b103ddd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input data: df_exp, df_score, df_outlier, pcc_tri, pcc\n",
    "# cts = No normalize\n",
    "# log = log transform\n",
    "# use log(cts) or log(cpm) compute Pearsons' correlation as global coexpression network edges.\n",
    "# expression table using raw counts log transformed (Optional also do count per million normalization)\n",
    "c, ctype, p = 'raw', 'cts', 'log' # for method develop purpose; leave them default\n",
    "\n",
    "thres_score = 0.0001 # threshold of ae scores\n",
    "thres_q_pcc = 0.01 # in percent threshold of whole pearsons matrix \n",
    "ae_gene_max = 20  # num of top ae genes \n",
    "pcc_gene_max = 10  # num of correlated genes per ae gene; affect network size significantly\n",
    "pcc_gene_min = 5 # if not have any correlated genes, take some genes below threshold.\n",
    "\n",
    "kwargs = {'output_dir': outdir,\n",
    "          'df_exp': df_exp, # df\n",
    "          'df_score': df_score, # df\n",
    "          'df_outlier': df_outlier, # df\n",
    "          'pcc_tri': pcc_tri,# df\n",
    "          'pcc': pcc, # df\n",
    "          'c': c, \n",
    "          'ctype': ctype,\n",
    "          'p': p,\n",
    "          'thres_score': thres_score, \n",
    "          'thres_q_pcc': thres_q_pcc,\n",
    "          'pcc_gene_min': pcc_gene_min, \n",
    "          'pcc_gene_max': pcc_gene_max, \n",
    "          'ae_gene_max': ae_gene_max,\n",
    "          'file_e': file_e, # SWEET need absolute file path to compute sample weight\n",
    "         }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0a567bf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------------------------------\n",
    "# Part 1. construct Reference Network for datasets and all samples( ~ 0.5min )\n",
    "# ------------------------------\n",
    "\n",
    "G_int = functions.global_net(**kwargs)\n",
    "logging.info(\"Global coexpress network of all genes\")\n",
    "kwargs['G_int'] = G_int     # Update kwargs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "85ea7774-0fd4-41c2-a987-edb948c986d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                                                                                                                  | 0/24 [00:38<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "# ------------------------------\n",
    "# Part 2. Generate AE Network for One Sample ( ~ 1min per sample)\n",
    "# ------------------------------\n",
    "\n",
    "for s in tqdm(s_list):\n",
    "    logging.info(f\"{s} AE network start\")\n",
    "    nodes, edges = functions.ae_net(s, **kwargs)\n",
    "    nodes.to_csv(filename_nd_template.format(outdir=outdir, sample=s), sep='\\t', header=True, index=True)\n",
    "    edges.to_csv(filename_ed_template.format(outdir=outdir, sample=s), sep='\\t', header=True, index=True)\n",
    "    \n",
    "    # identifying aberrant edges\n",
    "    edges = functions.label_aberrant_edges(edges)\n",
    "    edges.to_csv(filename_ed_aberrant_template.format(outdir=outdir, sample=s), sep='\\t', header=True, index=True)\n",
    "    # scatter plots to show aberrant edges\n",
    "    fig, axs = functions.create_scatter_plot(edges)\n",
    "    fig.savefig(filename_scatter_template.format(outdir=outdir, sample=s ), bbox_inches='tight')\n",
    "    plt.close(fig)\n",
    "    logging.info(f\"{s} AE network finish\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9e6d79c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> /mnt/disk7t/xwj/axolotl_rev/output/log/ae_net_20250519_222056.log <==\n",
      "2025-05-19 22:21:34,145:INFO:prior\n",
      "0.5    42\n",
      "1.0    26\n",
      "Name: count, dtype: int64\n",
      "2025-05-19 22:21:34,146:INFO:AE net: check AE network connectivity, 2\n",
      "2025-05-19 22:21:34,146:INFO:Nodes:163      Edges:223      Is Connected:True\n",
      "\n",
      "2025-05-19 22:21:34,632:INFO:AE net: run SWEET analysis\n",
      "2025-05-19 22:21:37,246:INFO:patients:{'R60537'}, index:[252], total samples:269, total genes:12543\n",
      "\n",
      "\n",
      "==> /mnt/disk7t/xwj/axolotl_rev/output/log/ae_net_20250519_225237.log <==\n",
      "2025-05-19 22:53:15,315:INFO:Nodes:163      Edges:223      Is Connected:True\n",
      "\n",
      "2025-05-19 22:53:15,864:INFO:AE net: run SWEET analysis\n",
      "2025-05-19 22:53:18,483:INFO:patients:{'R60537'}, index:[252], total samples:269, total genes:12543\n",
      "\n",
      "2025-05-19 22:53:39,761:INFO:R60537 AE network finish\n",
      "2025-05-20 08:05:23,084:ERROR:No such comm target registered: jupyter.widget.control\n",
      "2025-05-20 08:05:23,088:WARNING:No such comm: 85a52789-9f04-4f76-8a54-0a364afee28a\n",
      "2025-05-20 08:29:30,626:ERROR:No such comm target registered: jupyter.widget.control\n",
      "2025-05-20 08:29:30,629:WARNING:No such comm: 5a691619-1b73-4cca-879f-d41fbc1b1cdc\n",
      "\n",
      "==> /mnt/disk7t/xwj/axolotl_rev/output/log/ae_net_20250520_100320.log <==\n",
      "0.5     67\n",
      "2.0      3\n",
      "Name: count, dtype: int64\n",
      "2025-05-20 10:03:58,772:INFO:AE net: check AE network connectivity, 2\n",
      "2025-05-20 10:03:58,773:INFO:Nodes:286      Edges:539      Is Connected:True\n",
      "\n",
      "2025-05-20 10:03:59,980:INFO:AE net: run SWEET analysis\n",
      "2025-05-20 10:04:01,129:INFO:patients:{'R80184'}, index:[95], total samples:154, total genes:12543\n",
      "\n",
      "2025-05-20 10:04:31,489:INFO:R80184 AE network finish\n",
      "\n",
      "==> /mnt/disk7t/xwj/axolotl_rev/output/log/ae_net.log <==\n",
      "2025-05-19 22:19:29,455:INFO:prior\n",
      "0.5    42\n",
      "1.0    26\n",
      "Name: count, dtype: int64\n",
      "2025-05-19 22:19:29,456:INFO:AE net: check AE network connectivity, 2\n",
      "2025-05-19 22:19:29,456:INFO:Nodes:160      Edges:220      Is Connected:True\n",
      "\n",
      "2025-05-19 22:19:30,006:INFO:AE net: run SWEET analysis\n",
      "2025-05-19 22:19:32,666:INFO:patients:{'R60537'}, index:[252], total samples:269, total genes:12543\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! tail '{outdir}/log/ae_net'*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "05d4890c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %run /public236T/test1/axolotl_rev/script/sweet_functions.py\n",
    "# %run /mnt/disk7t/xwj/axolotl_rev/script/functions.py"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3.12]",
   "language": "python",
   "name": "conda-env-py3.12-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
