{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8593d81e",
   "metadata": {},
   "source": [
    "read ae expanded & annotate network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d5527a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os, sys\n",
    "from tqdm import tqdm\n",
    "plt.style.use('default')\n",
    "\n",
    "# load ae_net source code\n",
    "sys.path.append('/mnt/disk7t/xwj/axolotl_rev/gitcode')\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0e043a8",
   "metadata": {},
   "source": [
    "load packaged functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9378116e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/disk7t/xwj/soft/miniforge3/envs/py3.12/bin/python\n"
     ]
    }
   ],
   "source": [
    "! which python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed1918c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# output directory\n",
    "outdir = '/mnt/disk7t/xwj/axolotl_rev/output' \n",
    "\n",
    "os.makedirs(outdir+'/data', exist_ok=True)\n",
    "os.makedirs(outdir+'/sweet', exist_ok=True)\n",
    "\n",
    "filename_nd_template = '{outdir}/data/node_df_{sample}.txt'\n",
    "filename_ed_template = '{outdir}/data/edge_df_{sample}.txt'\n",
    "filename_ed_aberrant_template = '{outdir}/data/edge_df_{sample}_aberrant.txt'\n",
    "filename_scatter_template = '{outdir}/data/edge_scatter_{sample}.pdf'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "850c3e07",
   "metadata": {},
   "source": [
    "reading and processing different datasets related to gene expression and outlier detection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "47c0da66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to files from upstream AE detection tools\n",
    "datadir = '/mnt/disk7t/xwj/axolotl_rev/pfib_423/'\n",
    "\n",
    "## example score of pfib423 dataset\n",
    "result_baseline = {\n",
    "    'outsingle':f'{datadir}/osg_res_1_001.txt',\n",
    "    'outrider': f'{datadir}/outrider_res_1_001.txt',\n",
    "    'outrider_cts': f'{datadir}/outrider_norm_cts_1_001.txt',\n",
    "    'trueoutlier': f'{datadir}/df_true49.txt',\n",
    "}\n",
    "# your_dataset_file = '/media/eys/xwj/axolotl_rev/pfib_423/df_cts_gm2022_dedup.txt'\n",
    "\n",
    "#------------------------------------------------------------------------#\n",
    "# 原文提供了全部样本的详细注释信息，可以用来对数据集进行去批次效应\n",
    "file = f'{datadir}/sample_annot.tsv'\n",
    "s_anno = pd.read_csv(file, sep=\"\\t\", index_col='RNA_ID')\n",
    "# pfib数据集包含两种测序平台strand & non strand，two groups 表现为两个批次. 分别保存\n",
    "# rename: stranded -> ss； non-stranded -> ns\n",
    "s_anno['group'] = s_anno['STRAND_SPECIFIC'].map({True:'ss', False:'ns'})\n",
    "s_anno.describe()\n",
    "\n",
    "#------------------------------------------------------------------------#\n",
    "'''上游异常表达基因识别工具的输出结果, 此处假设有2种方法的结果,分别为OUTSINGLE和OUTRIDER'''\n",
    "\n",
    "# method 1. \n",
    "file = result_baseline['outsingle']\n",
    "score = pd.read_csv(file, sep=\"\\t\", index_col=0)\n",
    "\n",
    "# method 2.\n",
    "# if use OUTRIDER res\n",
    "file = result_baseline['outrider']\n",
    "score = pd.read_csv(file, sep=\"\\t\", index_col=0)\n",
    "score = score.pivot_table(values='pValue',index= 'geneID', columns='sampleID')\n",
    "\n",
    "'''not used'''\n",
    "cts_outrider = pd.read_csv(result_baseline['outrider_cts'], sep=\"\\t\", index_col=0)\n",
    "\n",
    "#------------------------------------------------------------------------#\n",
    "# 根据原文附件，我整理了这组样本，每个样本有candidate异常基因, 作为数据集的label\n",
    "# file = f'{datadir}/df_true3.txt'\n",
    "# file = f'{datadir}/df_true49.txt'\n",
    "# outlier = pd.read_csv(file, sep=\"\\t\", index_col=None)\n",
    "outlier = pd.read_csv(result_baseline['trueoutlier'], sep=\"\\t\", index_col=None)\n",
    "outlier['group'] = s_anno.loc[ outlier['Sample'],'group'].values\n",
    "\n",
    "#------------------------------------------------------------------------#\n",
    "'''表达量矩阵，作为共表达相关性依据，用于寻找异常表达基因之间的最短路径'''\n",
    "file = f'{datadir}/t0_g12543_s423.txt.gz'\n",
    "cts = pd.read_csv(file, sep=\"\\t\", index_col=0)\n",
    "exp = cts.transform(lambda x: np.log(x+1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ba561d0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Gene</th>\n",
       "      <th>RNAdefects</th>\n",
       "      <th>TableName</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R80184</td>\n",
       "      <td>M</td>\n",
       "      <td>ALDH18A1</td>\n",
       "      <td>AE, MAE</td>\n",
       "      <td>S2</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R60537</td>\n",
       "      <td>M</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>AE, Var</td>\n",
       "      <td>S2</td>\n",
       "      <td>ss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R62943</td>\n",
       "      <td>F</td>\n",
       "      <td>MICOS13</td>\n",
       "      <td>AE, AS</td>\n",
       "      <td>S2</td>\n",
       "      <td>ss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R96820</td>\n",
       "      <td>F</td>\n",
       "      <td>CLPP</td>\n",
       "      <td>AE, AS</td>\n",
       "      <td>S2</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R77611</td>\n",
       "      <td>F</td>\n",
       "      <td>DLD</td>\n",
       "      <td>AE, MAE</td>\n",
       "      <td>S2</td>\n",
       "      <td>ss</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample Sex      Gene RNAdefects TableName group\n",
       "0  R80184   M  ALDH18A1    AE, MAE        S2    ns\n",
       "1  R60537   M   ATP6AP1    AE, Var        S2    ss\n",
       "2  R62943   F   MICOS13     AE, AS        S2    ss\n",
       "3  R96820   F      CLPP     AE, AS        S2    ns\n",
       "4  R77611   F       DLD    AE, MAE        S2    ss"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outlier.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6d1b6152",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDIVIDUAL_ID</th>\n",
       "      <th>TISSUE</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AFFECTED</th>\n",
       "      <th>ICD_10</th>\n",
       "      <th>PAIRED_END</th>\n",
       "      <th>STRAND_SPECIFIC</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R0052</th>\n",
       "      <td>p11688</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Female</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R0053</th>\n",
       "      <td>p48517</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Male</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R19907</th>\n",
       "      <td>p46808</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Male</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R10581</th>\n",
       "      <td>p51250</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Female</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R0055</th>\n",
       "      <td>p5413</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Female</td>\n",
       "      <td>True</td>\n",
       "      <td>E88</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ns</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       INDIVIDUAL_ID      TISSUE     SEX  AFFECTED ICD_10  PAIRED_END  \\\n",
       "RNA_ID                                                                  \n",
       "R0052         p11688  Fibroblast  Female      True    E88        True   \n",
       "R0053         p48517  Fibroblast    Male      True    E88        True   \n",
       "R19907        p46808  Fibroblast    Male      True    E88        True   \n",
       "R10581        p51250  Fibroblast  Female      True    E88        True   \n",
       "R0055          p5413  Fibroblast  Female      True    E88        True   \n",
       "\n",
       "        STRAND_SPECIFIC group  \n",
       "RNA_ID                         \n",
       "R0052             False    ns  \n",
       "R0053             False    ns  \n",
       "R19907            False    ns  \n",
       "R10581            False    ns  \n",
       "R0055             False    ns  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_anno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c658dd1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ss 269 25\n",
      "ns 154 24\n",
      "{'ss': {'exp': '/mnt/disk7t/xwj/axolotl_rev/output/df_exp_ss269.txt', 'score': '/mnt/disk7t/xwj/axolotl_rev/output/df_score_ss269.txt', 'outlier': '/mnt/disk7t/xwj/axolotl_rev/output/df_true_ss25.txt'}, 'ns': {'exp': '/mnt/disk7t/xwj/axolotl_rev/output/df_exp_ns154.txt', 'score': '/mnt/disk7t/xwj/axolotl_rev/output/df_score_ns154.txt', 'outlier': '/mnt/disk7t/xwj/axolotl_rev/output/df_true_ns24.txt'}}\n"
     ]
    }
   ],
   "source": [
    "# dataset split by sequencing library type (stranded or non-stranded) as batches\n",
    "batches = ['ss','ns']\n",
    "\n",
    "# filename dict\n",
    "dict_filename = dict.fromkeys(batches)\n",
    "\n",
    "for group in batches:\n",
    "    \n",
    "    # expression matrix \n",
    "    sample_list = s_anno.query('group == @group').index.tolist()\n",
    "    expfile = f'{outdir}/df_exp_{group}{len(sample_list)}.txt'\n",
    "    exp.loc[:, sample_list].to_csv( expfile, sep='\\t')\n",
    "    \n",
    "    # score matrix\n",
    "    scorefile = f'{outdir}/df_score_{group}{len(sample_list)}.txt'\n",
    "    score.loc[:, sample_list].to_csv( scorefile, sep='\\t')\n",
    "    \n",
    "    # sample meta table\n",
    "    sample_list_interest = outlier.query('group == @group')['Sample']\n",
    "    outlierfile = f'{outdir}/df_true_{group}{len(sample_list_interest)}.txt'\n",
    "    outlier.query('Sample in @sample_list_interest').to_csv(outlierfile,sep='\\t')\n",
    "        \n",
    "    dict_filename[group] = {}\n",
    "    dict_filename[group]['exp'] = expfile\n",
    "    dict_filename[group]['score'] = scorefile\n",
    "    dict_filename[group]['outlier'] = outlierfile\n",
    "    print(group, len(sample_list), len(sample_list_interest))\n",
    "    \n",
    "print(dict_filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a331c3f4",
   "metadata": {},
   "source": [
    "### run aberrant network and edge analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86de9fb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STEP: load expression data...\n",
      "Current Time: 2025-05-18 22:06:07 Elapsed Time: 0:00:01.130880\n",
      "\n",
      "STEP: Prepare coexpression pcc matrix\n",
      "Current Time: 2025-05-18 22:06:10 Elapsed Time: 0:00:03.360219\n",
      "\n",
      "STEP: Prepare coexpression edge table\n",
      "Current Time: 2025-05-18 22:06:14 Elapsed Time: 0:00:08.021681\n",
      "\n",
      "DataFrame       | Shape     \n",
      "----------------------------------------\n",
      "df_outlier      | (25, 6)\n",
      "df_exp          | (12543, 269)\n",
      "df_score        | (12543, 269)\n",
      "pcc             | (12543, 12543)\n",
      "pcc_tri         | (78657153, 3)\n",
      "\n",
      "\n",
      "Nodes:8447     Edges:786572   Is Connected:False\n",
      "\n",
      "Nodes:12543    Edges:828197   Is Connected:True\n",
      "\n",
      "STEP: Global coexpress network of all genes\n",
      "Current Time: 2025-05-18 22:06:41 Elapsed Time: 0:00:35.210311\n",
      "\n",
      "CPU times: user 49.8 s, sys: 4.16 s, total: 53.9 s\n",
      "Wall time: 35.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "start_time = datetime.datetime.now()\n",
    "\n",
    "thres_score = 0.0001 # threshold of ae scores\n",
    "thres_q_pcc = 0.01 # in percent threshold of whole pearsons matrix \n",
    "ae_gene_max = 20  # num of top ae genes \n",
    "pcc_gene_max = 10  # num of correlated genes per ae gene; affect network size significantly\n",
    "pcc_gene_min = 5 # if not have any correlated genes, take some genes below threshold.\n",
    "\n",
    "# ------------------------------\n",
    "# Part 1. Dataset Batch & Sample Selection \n",
    "# ------------------------------\n",
    "# cts = No normalize\n",
    "# log = log transform\n",
    "# use log(cts) or log(cpm) compute Pearsons' correlation as global coexpression network edges.\n",
    "# expression table using raw counts log transformed (Optional also do count per million normalization)\n",
    "c, ctype, p = 'raw', 'cts', 'log' # user can leave it as default\n",
    "\n",
    "# s_list = [ 'R52016','R28774'] #from ns sample\n",
    "# s_list = [ \"R15264\" ] #from ss sample\n",
    "\n",
    "# load data\n",
    "group = 'ss'\n",
    "file_e = dict_filename[group]['exp']\n",
    "df_exp = pd.read_csv(file_e, sep=\"\\t\", index_col=0)\n",
    "\n",
    "file_sc = dict_filename[group]['score']\n",
    "df_score = pd.read_csv(file_sc, sep=\"\\t\", index_col=0)\n",
    "\n",
    "outlierfile = dict_filename[group]['outlier']\n",
    "df_outlier = pd.read_csv(outlierfile, sep=\"\\t\", index_col=0)\n",
    "\n",
    "# select subset of samples\n",
    "if True:\n",
    "    s_list = df_outlier['Sample'].tolist()\n",
    "\n",
    "log_time(start_time, \"load expression data...\")\n",
    "\n",
    "#-----------------------------------------------------------\n",
    "# prepare intermediate data\n",
    "#-----------------------------------------------------------\n",
    "\n",
    "# correlation\n",
    "pcc = pd.DataFrame( np.corrcoef(df_exp), index=df_exp.index.tolist(), columns=df_exp.index.tolist())\n",
    "pcc.index.name, pcc.columns.name ='gene1','gene2'\n",
    "\n",
    "log_time(start_time, \"Prepare coexpression pcc matrix\")\n",
    "\n",
    "# pcc remove duplicate, get undirected network = 0.5 min\n",
    "# pcc = dict_pcc[c][ctype][p].copy()\n",
    "# pcc_tri = pd.Series(pcc.stack()).reset_index().rename(columns={0:'pcc'})\n",
    "# pcc_tri = pcc_tri[ pcc_tri['gene1'] < pcc_tri['gene2'] ]\n",
    "\n",
    "pcc_tri = pearson_extract_upper_triangle(pcc)\n",
    "log_time(start_time, \"Prepare coexpression edge table\")\n",
    "\n",
    "print(f\"{'DataFrame':<15} | {'Shape':<10}\")\n",
    "print(\"-\" * 40)\n",
    "print(f\"{'df_outlier':<15} | {df_outlier.shape}\")\n",
    "print(f\"{'df_exp':<15} | {df_exp.shape}\")\n",
    "print(f\"{'df_score':<15} | {df_score.shape}\")\n",
    "print(f\"{'pcc':<15} | {pcc.shape}\")\n",
    "print(f\"{'pcc_tri':<15} | {pcc_tri.shape}\")\n",
    "print(\"\\n\")\n",
    "\n",
    "# input data: df_exp, df_score, df_outlier, pcc_tri, pcc\n",
    "kwargs = {'output_dir': outdir,\n",
    "          'df_exp': df_exp, # df\n",
    "          'df_score': df_score, # df\n",
    "          'df_outlier': df_outlier, # df\n",
    "          'pcc_tri': pcc_tri,# df\n",
    "          'pcc': pcc, # df\n",
    "          'c': c, \n",
    "          'ctype': ctype,\n",
    "          'p': p,\n",
    "          'thres_score': thres_score, \n",
    "          'thres_q_pcc': thres_q_pcc,\n",
    "          'pcc_gene_min': pcc_gene_min, \n",
    "          'pcc_gene_max': pcc_gene_max, \n",
    "          'ae_gene_max': ae_gene_max,\n",
    "          'file_e': file_e, # sweet need this file path to compute sample weight\n",
    "         }\n",
    "\n",
    "# ------------------------------\n",
    "# Part 1. construct Reference Network for datasets and all samples( ~ 0.5min )\n",
    "# ------------------------------\n",
    "G_int = global_net(**kwargs)\n",
    "log_time(start_time, \"Global coexpress network of all genes\")\n",
    "kwargs['G_int'] = G_int     # Update kwargs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11749527",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                                                                                                                  | 0/25 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STEP: R60537 AE network start\n",
      "Current Time: 2025-05-18 22:23:26 Elapsed Time: 0:17:20.202269\n",
      "\n",
      "AE net: get AE genes\n",
      "Gene       | AE Value   | Edge      \n",
      "----------------------------------------\n",
      "ATP6AP1    | 3.82e-40   | 7    \n",
      "SRI        | 6.02e-11   | 10   \n",
      "TP53I3     | 3.35e-09   | 5    \n",
      "ARMC10     | 1.56e-06   | 10   \n",
      "DAP3       | 2.53e-05   | 10   \n",
      "MRPL3      | 6.54e-05   | 10   \n",
      "MCM3       | 6.78e-05   | 10   \n",
      "SYNM       | 7.58e-05   | 5    \n",
      "CUTALP     | 8.48e-05   | 5    \n",
      "RPS6       | 9.08e-05   | 10   \n",
      "\n",
      "\n",
      "AE net: check AE network connectivity, 1\n",
      "Nodes:92       Edges:82       Is Connected:False\n",
      "\n",
      "Modules: [8, 11, 6, 11, 11, 11, 11, 6, 6, 11]\n",
      "AE net: shortest paths between modules\n",
      " prior\n",
      "1.0    73\n",
      "0.5    14\n",
      "2.0    11\n",
      "Name: count, dtype: int64\n",
      "AE net: check AE network connectivity, 2\n",
      "Nodes:158      Edges:246      Is Connected:True\n",
      "\n",
      "AE net: run SWEET analysis\n",
      "patients:{'R60537'}, index:[252], total samples:269, total genes:12543\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|██████▊                                                                                                                                                                   | 1/25 [00:30<12:18, 30.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STEP: R60537 AE network finish\n",
      "Current Time: 2025-05-18 22:23:57 Elapsed Time: 0:17:50.980084\n",
      "\n",
      "STEP: R62943 AE network start\n",
      "Current Time: 2025-05-18 22:23:57 Elapsed Time: 0:17:50.982713\n",
      "\n",
      "AE net: get AE genes\n",
      "Gene       | AE Value   | Edge      \n",
      "----------------------------------------\n",
      "MICOS13    | 3.50e-19   | 10   \n",
      "ATIC       | 7.81e-17   | 10   \n",
      "SUOX       | 3.20e-12   | 6    \n",
      "PXDC1      | 7.29e-06   | 5    \n",
      "SEPTIN2    | 1.95e-05   | 10   \n",
      "BMS1P5     | 4.45e-05   | 5    \n",
      "KHDRBS3    | 9.80e-05   | 5    \n",
      "\n",
      "\n",
      "AE net: check AE network connectivity, 1\n",
      "Nodes:58       Edges:51       Is Connected:False\n",
      "\n",
      "Modules: [11, 11, 7, 6, 11, 6, 6]\n",
      "AE net: shortest paths between modules\n",
      " prior\n",
      "1.0    35\n",
      "0.5     7\n",
      "2.0     2\n",
      "Name: count, dtype: int64\n",
      "AE net: check AE network connectivity, 2\n",
      "Nodes:109      Edges:148      Is Connected:True\n",
      "\n",
      "AE net: run SWEET analysis\n",
      "patients:{'R62943'}, index:[14], total samples:269, total genes:12543\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  8%|█████████████▌                                                                                                                                                            | 2/25 [00:58<11:11, 29.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STEP: R62943 AE network finish\n",
      "Current Time: 2025-05-18 22:24:25 Elapsed Time: 0:18:19.059665\n",
      "\n",
      "STEP: R77611 AE network start\n",
      "Current Time: 2025-05-18 22:24:25 Elapsed Time: 0:18:19.061050\n",
      "\n",
      "AE net: get AE genes\n",
      "Gene       | AE Value   | Edge      \n",
      "----------------------------------------\n",
      "DLD        | 5.75e-30   | 10   \n",
      "PUM3       | 1.20e-16   | 10   \n",
      "DDX24      | 2.17e-09   | 10   \n",
      "PIK3C2A    | 1.72e-08   | 10   \n",
      "NRBP2      | 2.18e-08   | 5    \n",
      "B3GALNT2   | 4.03e-08   | 10   \n",
      "DOCK9      | 1.38e-07   | 5    \n",
      "UGGT2      | 4.73e-06   | 10   \n",
      "RBL1       | 1.16e-05   | 10   \n",
      "KRT34      | 2.51e-05   | 5    \n",
      "SRP72      | 3.76e-05   | 10   \n",
      "DPH6       | 5.16e-05   | 10   \n",
      "PSMG3-AS1  | 7.01e-05   | 5    \n",
      "NUBPL      | 7.90e-05   | 10   \n",
      "\n",
      "\n",
      "AE net: check AE network connectivity, 1\n",
      "Nodes:129      Edges:120      Is Connected:False\n",
      "\n",
      "Modules: [39, 11, 11, 11, 6, 11, 6, 11, 11, 6, 6]\n",
      "AE net: shortest paths between modules\n",
      " prior\n",
      "1.0    135\n",
      "0.5     17\n",
      "2.0     12\n",
      "Name: count, dtype: int64\n",
      "AE net: check AE network connectivity, 2\n",
      "Nodes:270      Edges:436      Is Connected:True\n",
      "\n",
      "AE net: run SWEET analysis\n",
      "patients:{'R77611'}, index:[160], total samples:269, total genes:12543\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# ------------------------------\n",
    "# Part 2. Generate AE Network for One Sample\n",
    "# ------------------------------\n",
    "\n",
    "\n",
    "for s in tqdm(s_list):\n",
    "    \n",
    "    log_time(start_time, f\"{s} AE network start\")\n",
    "    \n",
    "    nodes, edges = ae_net(s, **kwargs)\n",
    "    nodes.to_csv(filename_nd_template.format(outdir=outdir, sample=s), sep='\\t', header=True, index=True)\n",
    "    edges.to_csv(filename_ed_template.format(outdir=outdir, sample=s), sep='\\t', header=True, index=True)\n",
    "    \n",
    "    # identifying aberrant edges\n",
    "    edges = label_aberrant_edges(edges)\n",
    "    edges.to_csv(filename_ed_aberrant_template.format(outdir=outdir, sample=s), sep='\\t', header=True, index=True)\n",
    "    # scatter plots to show aberrant edges\n",
    "    fig, axs = create_scatter_plot(edges)\n",
    "    fig.savefig(filename_scatter_template.format(outdir=outdir, sample=s ), bbox_inches='tight')\n",
    "    plt.close(fig)\n",
    "    \n",
    "    log_time(start_time, f\"{s} AE network finish\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96c089dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! which python\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c52166dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# label aberrant edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "880a49a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "    # plot edge attributes and labels\n",
    "    # n_row, n_col = 1, 3\n",
    "    # fig, axs = plt.subplots(n_row, n_col, sharey=False,figsize=(6*n_col, 6*n_row))\n",
    "    # plt.subplots_adjust( wspace=0.2, hspace=0.2)\n",
    "    # sns.scatterplot(data=edges, x = 'residual_z', y = 'outscore', hue='group', style='label', size='abs(sweet-pcc)', alpha=0.8, ax=axs[0])\n",
    "    # sns.scatterplot(data=edges, x = 'residual_z', y = 'abs(sweet-pcc)', hue='group', style='label', size='outscore', alpha=0.8,ax=axs[1])\n",
    "    # sns.scatterplot(data=edges, x = 'abs(sweet-pcc)', y = 'outscore', hue='group', style='label', alpha=0.8, ax=axs[2])\n",
    "    # # idx = 3\n",
    "    # # sns.scatterplot(data=edges, x = 'residual_z', y = 'abs(sin-pcc)', hue='group', size='outscore', alpha=0.8,ax=axs[idx])\n",
    "    # plt.savefig(filename_scatter_template.format(outdir=datadir, sample=s ), bbox_inches='tight')\n",
    "    # plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b976fb5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>group</th>\n",
       "      <th>sweet</th>\n",
       "      <th>pcc</th>\n",
       "      <th>outscore</th>\n",
       "      <th>residual_z</th>\n",
       "      <th>abs(sweet-pcc)</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TANGO2</td>\n",
       "      <td>DCTN1</td>\n",
       "      <td>AE</td>\n",
       "      <td>0.4646</td>\n",
       "      <td>0.838919</td>\n",
       "      <td>3.533572</td>\n",
       "      <td>-7.046113</td>\n",
       "      <td>0.374319</td>\n",
       "      <td>aberrant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TANGO2</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>AE</td>\n",
       "      <td>0.2751</td>\n",
       "      <td>0.836316</td>\n",
       "      <td>3.838883</td>\n",
       "      <td>-7.901387</td>\n",
       "      <td>0.561170</td>\n",
       "      <td>aberrant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TANGO2</td>\n",
       "      <td>TBC1D13</td>\n",
       "      <td>AE</td>\n",
       "      <td>0.4954</td>\n",
       "      <td>0.832722</td>\n",
       "      <td>3.442074</td>\n",
       "      <td>-6.774214</td>\n",
       "      <td>0.337360</td>\n",
       "      <td>aberrant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TANGO2</td>\n",
       "      <td>WIPI2</td>\n",
       "      <td>AE</td>\n",
       "      <td>0.4563</td>\n",
       "      <td>0.830722</td>\n",
       "      <td>3.707770</td>\n",
       "      <td>-6.936302</td>\n",
       "      <td>0.374423</td>\n",
       "      <td>aberrant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TANGO2</td>\n",
       "      <td>RAB5C</td>\n",
       "      <td>AE</td>\n",
       "      <td>0.4500</td>\n",
       "      <td>0.827429</td>\n",
       "      <td>3.755403</td>\n",
       "      <td>-6.909761</td>\n",
       "      <td>0.377478</td>\n",
       "      <td>aberrant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2208</th>\n",
       "      <td>TOX4</td>\n",
       "      <td>FAF2</td>\n",
       "      <td>ADD2</td>\n",
       "      <td>0.8755</td>\n",
       "      <td>0.901236</td>\n",
       "      <td>1.647343</td>\n",
       "      <td>1.731929</td>\n",
       "      <td>0.025747</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2209</th>\n",
       "      <td>PBXIP1</td>\n",
       "      <td>BBS1</td>\n",
       "      <td>ADD2</td>\n",
       "      <td>0.8696</td>\n",
       "      <td>0.869819</td>\n",
       "      <td>1.011160</td>\n",
       "      <td>-0.136087</td>\n",
       "      <td>0.000190</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2210</th>\n",
       "      <td>PBXIP1</td>\n",
       "      <td>MEGF8</td>\n",
       "      <td>ADD2</td>\n",
       "      <td>0.8706</td>\n",
       "      <td>0.870489</td>\n",
       "      <td>1.040907</td>\n",
       "      <td>0.121240</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2211</th>\n",
       "      <td>CCM2</td>\n",
       "      <td>PPP1CA</td>\n",
       "      <td>ADD2</td>\n",
       "      <td>0.9040</td>\n",
       "      <td>0.898310</td>\n",
       "      <td>1.020430</td>\n",
       "      <td>-0.002923</td>\n",
       "      <td>0.005498</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2212</th>\n",
       "      <td>BBS1</td>\n",
       "      <td>SHFL</td>\n",
       "      <td>ADD2</td>\n",
       "      <td>0.8520</td>\n",
       "      <td>0.855592</td>\n",
       "      <td>1.019081</td>\n",
       "      <td>0.510997</td>\n",
       "      <td>0.003541</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2213 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene1    gene2 group   sweet       pcc  outscore  residual_z  \\\n",
       "0     TANGO2    DCTN1    AE  0.4646  0.838919  3.533572   -7.046113   \n",
       "1     TANGO2  TMEM127    AE  0.2751  0.836316  3.838883   -7.901387   \n",
       "2     TANGO2  TBC1D13    AE  0.4954  0.832722  3.442074   -6.774214   \n",
       "3     TANGO2    WIPI2    AE  0.4563  0.830722  3.707770   -6.936302   \n",
       "4     TANGO2    RAB5C    AE  0.4500  0.827429  3.755403   -6.909761   \n",
       "...      ...      ...   ...     ...       ...       ...         ...   \n",
       "2208    TOX4     FAF2  ADD2  0.8755  0.901236  1.647343    1.731929   \n",
       "2209  PBXIP1     BBS1  ADD2  0.8696  0.869819  1.011160   -0.136087   \n",
       "2210  PBXIP1    MEGF8  ADD2  0.8706  0.870489  1.040907    0.121240   \n",
       "2211    CCM2   PPP1CA  ADD2  0.9040  0.898310  1.020430   -0.002923   \n",
       "2212    BBS1     SHFL  ADD2  0.8520  0.855592  1.019081    0.510997   \n",
       "\n",
       "      abs(sweet-pcc)     label  \n",
       "0           0.374319  aberrant  \n",
       "1           0.561170  aberrant  \n",
       "2           0.337360  aberrant  \n",
       "3           0.374423  aberrant  \n",
       "4           0.377478  aberrant  \n",
       "...              ...       ...  \n",
       "2208        0.025747    normal  \n",
       "2209        0.000190    normal  \n",
       "2210        0.000116    normal  \n",
       "2211        0.005498    normal  \n",
       "2212        0.003541    normal  \n",
       "\n",
       "[2213 rows x 9 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3.12]",
   "language": "python",
   "name": "conda-env-py3.12-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
